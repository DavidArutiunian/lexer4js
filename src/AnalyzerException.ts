import Nullable from "./Nullable";
import WordStateFactory, { WordState } from "./WordState";

const WordState = WordStateFactory();

export class AnalyzerException extends Error {
  constructor(source: string, position: number) {
    super("");

    const substring = source
      .slice(
        position - AnalyzerException.getStartOffsetIndex(source, position),
        position + AnalyzerException.getEndOffsetIndex(source, position)
      )
      .trim();

    this.message = `You have an error in your syntax near ${substring}`;
  }

  private static getStartOffsetIndex(source: string, position: number): number {
    let index = 0;
    let state: Nullable<WordState> = WordState.INITIAL;

    while (position - index !== 0 && state?.whitespaces !== WordState.WORD_LIMIT && state !== WordState.NEWLINE) {
      index++;

      const char = source.charAt(position + index);
      const property = state?.next(char);

      if (property) {
        state = WordState[property];
      }
    }

    WordState.reset();

    return index;
  }

  private static getEndOffsetIndex(source: string, position: number): number {
    let index = 0;
    let state = WordState.INITIAL;

    while (
      position + index !== source.length - 1 &&
      state.whitespaces !== WordState.WORD_LIMIT &&
      state !== WordState.NEWLINE
    ) {
      index++;

      const char = source.charAt(position + index);
      const property = state?.next(char);

      if (property) {
        state = WordState[property];
      }
    }

    WordState.reset();

    return index;
  }
}
