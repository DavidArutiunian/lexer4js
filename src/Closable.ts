export interface Closable {
  close(): void;
}

export default Closable;
