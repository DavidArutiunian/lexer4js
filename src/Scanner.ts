export default class LineScanner {
  private position = 0;
  private lines: string[] = [];

  constructor(source: string) {
    this.lines = source.split("\n");
  }

  hasNextLine(): boolean {
    return this.position < this.lines.length;
  }

  readNextLine(): string {
    return this.lines[this.position++];
  }
}
