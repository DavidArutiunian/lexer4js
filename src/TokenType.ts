export enum TokenType {
  BLOCK_COMMENT = "BLOCK_COMMENT",
  BLOCK_COMMENT_UNTERMINATED = "BLOCK_COMMENT_UNTERMINATED",
  LINE_COMMENT = "LINE_COMMENT",
  WHITE_SPACE = "WHITE_SPACE",
  TAB = "TAB",
  NEW_LINE = "NEW_LINE",
  CLOSING_BRACE = "CLOSING_BRACE",
  OPENING_BRACE = "OPENING_BRACE",
  OPENING_CURLY_BRACE = "OPENING_CURLY_BRACE",
  CLOSING_CURLY_BRACE = "CLOSING_CURLY_BRACE",
  SCIENTIFIC_LITERAL = "SCIENTIFIC_LITERAL",
  OCTAL_LITERAL = "OCTAL_LITERAL",
  DOUBLE_LITERAL = "DOUBLE_LITERAL",
  // FLOAT_LITERAL = "FLOAT_LITERAL",
  INT_LITERAL = "INT_LITERAL",
  HEX_LITERAL = "HEX_LITERAL",
  BINARY_LITERAL = "BINARY_LITERAL",
  ADDITION = "ADDITION",
  SUBTRACTION = "SUBTRACTION",
  MULTIPLICATION = "MULTIPLICATION",
  DIVISION = "DIVISION",
  XOR = "XOR",
  POINT = "POINT",
  EQUALS = "EQUALS",
  NOT_EQUALS = "NOT_EQUALS",
  ASSIGNMENT = "ASSIGNMENT",
  MORE = "MORE",
  LESS = "LESS",
  AND = "AND",
  OR = "OR",
  STATIC = "STATIC",
  PUBLIC = "PUBLIC",
  PRIVATE = "PRIVATE",
  INT = "INT",
  CHAR = "CHAR",
  STRING = "STRING",
  DOUBLE = "DOUBLE",
  // FLOAT = "FLOAT",
  BOOLEAN = "BOOLEAN",
  VOID = "VOID",
  FALSE = "FALSE",
  TRUE = "TRUE",
  NULL = "NULL",
  RETURN = "RETURN",
  NEW = "NEW",
  CLASS = "CLASS",
  IF = "IF",
  WHILE = "WHILE",
  FOR = "FOR",
  ELSE = "ELSE",
  SEMICOLON = "SEMICOLON",
  COLON = "COLON",
  COMMA = "COMMA",
  IDENTIFIER = "IDENTIFIER",
  OPENING_SQUARE_BRACE = "OPENING_SQUARE_BRACE",
  CLOSING_SQUARE_BRACE = "CLOSING_SQUARE_BRACE",
  VAR = "VAR",
  FINAL = "FINAL",
  MULTILINE_STRING_LITERAL = "MULTILINE_STRING_LITERAL",
  STRING_LITERAL = "STRING_LITERAL",
  CHAR_LITERAL = "CHAR_LITERAL",
}

export function isTokenAuxiliary(token: TokenType): boolean {
  return (
    token == TokenType.BLOCK_COMMENT ||
    token == TokenType.LINE_COMMENT ||
    token == TokenType.NEW_LINE ||
    token == TokenType.TAB ||
    token == TokenType.WHITE_SPACE
  );
}

export default TokenType;
