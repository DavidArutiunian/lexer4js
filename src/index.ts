export { default as Lexer } from "./Lexer";
export { default as Token } from "./Token";
export { default as TokenType } from "./TokenType";
