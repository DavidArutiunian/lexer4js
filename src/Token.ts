import TokenType from "./TokenType";

export interface Token {
  literal: string;
  type: TokenType;
  line: number;
  position: number;
}

export default Token;
