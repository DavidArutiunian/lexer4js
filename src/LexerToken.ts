import Token from "./Token";
import TokenType, { isTokenAuxiliary } from "./TokenType";

export default class LexerToken implements Token {
  constructor(
    private to: number,
    public literal: string,
    public type: TokenType,
    public line: number,
    public position: number
  ) {}

  get endPosition(): number {
    return this.to;
  }

  get isNotAuxiliary(): boolean {
    return !isTokenAuxiliary(this.type);
  }

  toString(): string {
    if (this.isNotAuxiliary) {
      // prettier-ignore
      return this.type + " '" + this.literal.trim() + "' [L" + this.line + ":" + this.position + "]";
    }
    return this.type + " [L" + this.line + ":" + this.position + "]";
  }
}
