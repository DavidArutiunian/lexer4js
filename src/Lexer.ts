import AnalyzerException from "./AnalyzerException";
import LexerToken from "./LexerToken";
import Nullable from "./Nullable";
import LineScanner from "./Scanner";
import Token from "./Token";
import TokenType from "./TokenType";

export class Lexer {
  // Order matters! ðŸ˜°
  private regex: Record<TokenType, string> = {
    [TokenType.BLOCK_COMMENT]: "(/\\*.*?\\*/).*",
    [TokenType.BLOCK_COMMENT_UNTERMINATED]: "(/\\*.*?).*",
    [TokenType.LINE_COMMENT]: "(//(.*?)[\r$]?\n).*",
    [TokenType.WHITE_SPACE]: "( ).*",
    [TokenType.OPENING_BRACE]: "(\\().*",
    [TokenType.CLOSING_BRACE]: "(\\)).*",
    [TokenType.SEMICOLON]: "(;).*",
    [TokenType.COLON]: "(:).*",
    [TokenType.COMMA]: "(,).*",
    [TokenType.OPENING_CURLY_BRACE]: "(\\{).*",
    [TokenType.CLOSING_CURLY_BRACE]: "(\\}).*",
    [TokenType.SCIENTIFIC_LITERAL]: "\\b([+\\-]?(?:0|[1-9]\\d*)(?:\\.\\d*)?(?:[eE][+\\-]?\\d+))\\b.*",
    [TokenType.OCTAL_LITERAL]: "\\b(0[0-7]+)\\b.*",
    // [TokenType.FLOAT_LITERAL]: "\\b(\\d{1,9}\\.\\d{1,16})\\b.*",
    [TokenType.DOUBLE_LITERAL]: "\\b(\\d{1,9}\\.\\d{1,32})\\b.*",
    [TokenType.INT_LITERAL]: "\\b(\\d{1,9})\\b.*",
    [TokenType.HEX_LITERAL]: "\\b(0[xX][0-9a-fA-F]+)\\b.*",
    [TokenType.BINARY_LITERAL]: "\\b(0[bB][01]+)\\b.*",
    [TokenType.VOID]: "\\b(void)\\b.*",
    [TokenType.INT]: "\\b(int)\\b.*",
    [TokenType.CHAR]: "\\b(char)\\b.*",
    [TokenType.STRING]: "\\b(string)\\b.*",
    [TokenType.DOUBLE]: "\\b(double)\\b.*",
    // [TokenType.FLOAT]: "\\b(float)\\b.*",
    [TokenType.BOOLEAN]: "\\b(boolean)\\b.*",
    [TokenType.TAB]: "(\\t).*",
    [TokenType.NEW_LINE]: "(\\n).*",
    [TokenType.PUBLIC]: "\\b(public)\\b.*",
    [TokenType.PRIVATE]: "\\b(private)\\b.*",
    [TokenType.FALSE]: "\\b(false)\\b.*",
    [TokenType.TRUE]: "\\b(true)\\b.*",
    [TokenType.NULL]: "\\b(null)\\b.*",
    [TokenType.RETURN]: "\\b(return)\\b.*",
    [TokenType.NEW]: "\\b(new)\\b.*",
    [TokenType.CLASS]: "\\b(class)\\b.*",
    [TokenType.IF]: "\\b(if)\\b.*",
    [TokenType.ELSE]: "\\b(else)\\b.*",
    [TokenType.WHILE]: "\\b(while)\\b.*",
    [TokenType.FOR]: "\\b(for)\\b.*",
    [TokenType.STATIC]: "\\b(static)\\b.*",
    [TokenType.POINT]: "(\\.).*",
    [TokenType.ADDITION]: "(\\+{1}).*",
    [TokenType.SUBTRACTION]: "(\\-{1}).*",
    [TokenType.MULTIPLICATION]: "(\\*).*",
    [TokenType.DIVISION]: "(/).*",
    [TokenType.EQUALS]: "(==).*",
    [TokenType.XOR]: "(\\^).*",
    [TokenType.ASSIGNMENT]: "(=).*",
    [TokenType.NOT_EQUALS]: "(\\!=).*",
    [TokenType.MORE]: "(>).*",
    [TokenType.LESS]: "(<).*",
    [TokenType.AND]: "(&&).*",
    [TokenType.OR]: "(\\|\\|).*",
    [TokenType.IDENTIFIER]: "\\b([a-zA-Z]{1}[0-9a-zA-Z_]{0,31})\\b.*",
    [TokenType.OPENING_SQUARE_BRACE]: "(\\[).*",
    [TokenType.CLOSING_SQUARE_BRACE]: "(\\]).*",
    [TokenType.VAR]: "\\b(var)\\b.*",
    [TokenType.FINAL]: "\\b(final)\\b.*",
    [TokenType.STRING_LITERAL]: '(\\"([^\\\\\\"]|\\\\.)*\\").*',
    [TokenType.MULTILINE_STRING_LITERAL]: '(\\"\\"\\"([^\\\\\\"]|\\\\.)*\\"\\"\\").*',
    [TokenType.CHAR_LITERAL]: "('(.{1}|\\\\n|\\\\t)').*",
  };

  private result: LexerToken[] = [];

  public tokenize(source: string): Token[] {
    let position = 0;
    let token: Nullable<LexerToken> = null;
    do {
      token = this.getNextToken(source, position);
      // eslint-disable-next-line security/detect-possible-timing-attacks
      if (token !== null) {
        position = token.endPosition;
        this.result.push(token);
      }
    } while (token !== null && position !== source.length);
    if (position !== source.length) {
      throw new AnalyzerException(source, position);
    }
    return this.getFilteredTokens();
  }

  private getFilteredTokens(): Token[] {
    return this.result.filter((token) => token.isNotAuxiliary);
  }

  private getNextToken(source: string, from: number): Nullable<LexerToken> {
    if (from < 0 || from >= source.length) {
      throw new Error("Illegal index in the input stream");
    }
    for (const [type, regex] of Object.entries(this.regex)) {
      // eslint-disable-next-line security/detect-non-literal-regexp
      const pattern = new RegExp(`^${regex}$`, "s");
      const match = source.slice(from).match(pattern);
      if (match) {
        if (type === TokenType.BLOCK_COMMENT_UNTERMINATED) {
          return null;
        }
        const literal = match[1];
        const line = this.getLineNumber(source, from);
        const position = this.getPositionInLine(source, line, from);
        return new LexerToken(from + literal.length, literal, type as TokenType, line, position);
      }
    }
    return null;
  }

  private getLineNumber(source: string, from: number): number {
    return source.slice(0, from + 1).split("\n").length;
  }

  private getPositionInLine(source: string, line: number, from: number): number {
    const scanner = new LineScanner(source);
    let index = 0;
    let pos = from;
    while (scanner.hasNextLine() && index !== line - 1) {
      const skipped = scanner.readNextLine();
      pos -= skipped.length + 1;
      index++;
    }
    return pos;
  }
}

export default Lexer;
