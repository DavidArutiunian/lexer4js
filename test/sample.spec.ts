import fs from "fs";

import { Lexer } from "../src";

describe("Lexer", () => {
  const source = fs.readFileSync("test/code.java", "utf8");

  it("should match snapshot", () => {
    const lexer = new Lexer();
    const tokens = lexer.tokenize(source);
    const actual = tokens.map((token) => token.toString()).join("\n");

    expect(actual).toMatchSnapshot();
  });
});
