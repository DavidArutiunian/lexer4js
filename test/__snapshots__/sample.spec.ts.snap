// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Lexer should match snapshot 1`] = `
"CLASS 'class' [L1:0]
IDENTIFIER 'Foo' [L1:6]
OPENING_CURLY_BRACE '{' [L1:10]
PRIVATE 'private' [L2:4]
DOUBLE 'double' [L2:12]
IDENTIFIER 'big' [L2:19]
ASSIGNMENT '=' [L2:23]
SCIENTIFIC_LITERAL '3.2e+23' [L2:25]
SEMICOLON ';' [L2:32]
PRIVATE 'private' [L3:4]
DOUBLE 'double' [L3:12]
IDENTIFIER 'small' [L3:19]
ASSIGNMENT '=' [L3:25]
SUBTRACTION '-' [L3:27]
SCIENTIFIC_LITERAL '4.70e-9' [L3:28]
SEMICOLON ';' [L3:35]
PRIVATE 'private' [L5:4]
STRING 'string' [L5:12]
IDENTIFIER 'message' [L5:19]
ASSIGNMENT '=' [L5:27]
STRING_LITERAL '\\"FooBarBaz\\"' [L5:29]
SEMICOLON ';' [L5:40]
PRIVATE 'private' [L6:4]
CHAR 'char' [L6:12]
IDENTIFIER 'newline' [L6:17]
ASSIGNMENT '=' [L6:25]
CHAR_LITERAL ''\\\\n'' [L6:27]
SEMICOLON ';' [L6:31]
PRIVATE 'private' [L8:4]
INT 'int' [L8:12]
IDENTIFIER 'hex' [L8:16]
ASSIGNMENT '=' [L8:20]
HEX_LITERAL '0x0A0B0C' [L8:22]
SEMICOLON ';' [L8:30]
PRIVATE 'private' [L9:4]
INT 'int' [L9:12]
IDENTIFIER 'octal' [L9:16]
ASSIGNMENT '=' [L9:22]
OCTAL_LITERAL '0737' [L9:24]
SEMICOLON ';' [L9:28]
PRIVATE 'private' [L10:4]
INT 'int' [L10:12]
IDENTIFIER 'binary' [L10:16]
ASSIGNMENT '=' [L10:23]
BINARY_LITERAL '0b01001001110' [L10:25]
SEMICOLON ';' [L10:38]
PRIVATE 'private' [L12:4]
BOOLEAN 'boolean' [L12:12]
IDENTIFIER 'flag' [L12:20]
ASSIGNMENT '=' [L12:25]
FALSE 'false' [L12:27]
SEMICOLON ';' [L12:32]
PRIVATE 'private' [L14:4]
STRING 'string' [L14:12]
IDENTIFIER 'multiline' [L14:19]
ASSIGNMENT '=' [L14:29]
STRING_LITERAL '\\"\\"' [L14:31]
STRING_LITERAL '\\"
        Hello, World!
        Who I am?
    \\"' [L14:33]
STRING_LITERAL '\\"\\"' [L17:5]
SEMICOLON ';' [L17:7]
PUBLIC 'public' [L23:4]
VOID 'void' [L23:11]
IDENTIFIER 'main' [L23:16]
OPENING_BRACE '(' [L23:20]
STRING 'string' [L23:21]
OPENING_SQUARE_BRACE '[' [L23:27]
CLOSING_SQUARE_BRACE ']' [L23:28]
IDENTIFIER 'args' [L23:30]
CLOSING_BRACE ')' [L23:34]
OPENING_CURLY_BRACE '{' [L23:36]
INT 'int' [L24:8]
IDENTIFIER 'size' [L24:12]
ASSIGNMENT '=' [L24:17]
INT_LITERAL '3' [L24:19]
SEMICOLON ';' [L24:20]
INT 'int' [L25:8]
OPENING_SQUARE_BRACE '[' [L25:12]
IDENTIFIER 'size' [L25:13]
CLOSING_SQUARE_BRACE ']' [L25:17]
IDENTIFIER 'array' [L25:19]
ASSIGNMENT '=' [L25:25]
OPENING_CURLY_BRACE '{' [L25:27]
INT_LITERAL '1' [L25:29]
COMMA ',' [L25:30]
INT_LITERAL '2' [L25:32]
COMMA ',' [L25:33]
INT_LITERAL '3' [L25:35]
CLOSING_CURLY_BRACE '}' [L25:37]
SEMICOLON ';' [L25:38]
INT 'int' [L26:8]
IDENTIFIER 'index' [L26:12]
ASSIGNMENT '=' [L26:18]
INT_LITERAL '0' [L26:20]
SEMICOLON ';' [L26:21]
FLOAT 'float' [L27:8]
IDENTIFIER 'e' [L27:14]
ASSIGNMENT '=' [L27:16]
FLOAT_LITERAL '2.73' [L27:18]
SEMICOLON ';' [L27:22]
WHILE 'while' [L28:8]
OPENING_BRACE '(' [L28:14]
IDENTIFIER 'index' [L28:15]
NOT_EQUALS '!=' [L28:21]
INT_LITERAL '0' [L28:24]
CLOSING_BRACE ')' [L28:25]
OPENING_CURLY_BRACE '{' [L28:27]
IDENTIFIER 'index' [L29:12]
ASSIGNMENT '=' [L29:18]
IDENTIFIER 'index' [L29:20]
SUBTRACTION '-' [L29:26]
INT_LITERAL '1' [L29:28]
SEMICOLON ';' [L29:29]
IDENTIFIER 'var' [L30:12]
IDENTIFIER 'coefficient' [L30:16]
ASSIGNMENT '=' [L30:28]
IDENTIFIER 'big' [L30:30]
MULTIPLICATION '*' [L30:34]
IDENTIFIER 'small' [L30:36]
DIVISION '/' [L30:42]
IDENTIFIER 'hex' [L30:44]
SEMICOLON ';' [L30:47]
IDENTIFIER 'println' [L31:12]
OPENING_BRACE '(' [L31:19]
IDENTIFIER 'message' [L31:20]
COMMA ',' [L31:27]
IDENTIFIER 'array' [L31:29]
OPENING_SQUARE_BRACE '[' [L31:34]
IDENTIFIER 'index' [L31:35]
CLOSING_SQUARE_BRACE ']' [L31:40]
MULTIPLICATION '*' [L31:42]
IDENTIFIER 'coefficient' [L31:44]
COMMA ',' [L31:55]
IDENTIFIER 'newline' [L31:57]
CLOSING_BRACE ')' [L31:64]
SEMICOLON ';' [L31:65]
CLOSING_CURLY_BRACE '}' [L32:8]
FOR 'for' [L33:8]
OPENING_BRACE '(' [L33:12]
IDENTIFIER 'var' [L33:13]
IDENTIFIER 'num' [L33:17]
COLON ':' [L33:21]
IDENTIFIER 'array' [L33:23]
CLOSING_BRACE ')' [L33:28]
OPENING_CURLY_BRACE '{' [L33:30]
IDENTIFIER 'var' [L34:12]
IDENTIFIER 'coefficient' [L34:16]
ASSIGNMENT '=' [L34:28]
IDENTIFIER 'big' [L34:30]
MULTIPLICATION '*' [L34:34]
IDENTIFIER 'small' [L34:36]
DIVISION '/' [L34:42]
IDENTIFIER 'hex' [L34:44]
SEMICOLON ';' [L34:47]
IDENTIFIER 'println' [L35:12]
OPENING_BRACE '(' [L35:19]
IDENTIFIER 'message' [L35:20]
COMMA ',' [L35:27]
IDENTIFIER 'num' [L35:29]
MULTIPLICATION '*' [L35:33]
IDENTIFIER 'coefficient' [L35:35]
COMMA ',' [L35:46]
IDENTIFIER 'newline' [L35:48]
CLOSING_BRACE ')' [L35:55]
SEMICOLON ';' [L35:56]
CLOSING_CURLY_BRACE '}' [L36:8]
IDENTIFIER 'var' [L37:8]
IDENTIFIER 'secret' [L37:12]
ASSIGNMENT '=' [L37:19]
IDENTIFIER 'hex' [L37:21]
XOR '^' [L37:25]
IDENTIFIER 'octal' [L37:27]
XOR '^' [L37:33]
IDENTIFIER 'binary' [L37:35]
SEMICOLON ';' [L37:41]
IF 'if' [L38:8]
OPENING_BRACE '(' [L38:11]
IDENTIFIER 'flag' [L38:12]
AND '&&' [L38:17]
IDENTIFIER 'size' [L38:20]
OR '||' [L38:25]
IDENTIFIER 'secret' [L38:28]
CLOSING_BRACE ')' [L38:34]
OPENING_CURLY_BRACE '{' [L38:36]
IDENTIFIER 'println' [L39:10]
OPENING_BRACE '(' [L39:17]
IDENTIFIER 'secret' [L39:18]
XOR '^' [L39:25]
IDENTIFIER 'flag' [L39:27]
CLOSING_BRACE ')' [L39:31]
SEMICOLON ';' [L39:32]
CLOSING_CURLY_BRACE '}' [L40:8]
IDENTIFIER 'println' [L41:8]
OPENING_BRACE '(' [L41:15]
IDENTIFIER 'secret' [L41:16]
CLOSING_BRACE ')' [L41:22]
SEMICOLON ';' [L41:23]
CLOSING_CURLY_BRACE '}' [L42:4]
CLOSING_CURLY_BRACE '}' [L43:0]"
`;
